docs/content/specs/terraform/_index.md

```diff
@@ -533,6 +543,466 @@ Bad examples:
+      version = "3.11"
+    }
+  }
+}
+```
+
+Bad example:
+
+```terraform
+terraform {
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = ">= 3.11"
+    }
+  }
+}
+```
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR27 - Category: Code Style - Declaration of a provider in the module
+
+[By rules](https://www.terraform.io/docs/language/modules/develop/providers.html), in the module code `provider` cannot be declared. The only exception is when the module indeed need different instances of the same kind of `provider`(Eg. manipulating resources across different `location`s or accounts), you **MUST** declare `configuration_aliases` in `terraform.required_providers`. See details in this [document](https://www.terraform.io/docs/language/providers/configuration.html#alias-multiple-provider-configurations).
+
+`provider` block declared in the module can only be used to differentiate instances used in  `resource` and `data`. Declaration of fields other than `alias` in `provider` block is strictly forbidden. It could lead to module users unable to utilize `count`, `for_each` or `depends_on`. Configurations of the `provider` instance should be passed in by the module users.
+
+Good examples:
+
+In verified module:
+
+```terraform
+terraform {
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = "~> 3.0"
+      configuration_aliases = [ azurerm.alternate ]
+    }
+  }
+}
+```
+
+In the root module where we call this verified module:
+
+````terraform
+provider "azurerm" {
+  features {}
+}
+
+provider "azurerm" {
+  alias = "alternate"
+  features {}
+}
+
+module "foo" {
+  source = "xxx"
+  providers = {
+    azurerm = azurerm
+    azurerm.alternate = azurerm.alternate
+  }
+}
+````
+
+Bad example:
+
+In verified module:
+
+```terraform
+provider "azurerm" {
+  # Configuration options
+  features {}
+}
+```
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR28 - Category: Code Style - `output` **MUST** be arranged alphabetically
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR29 - Category: Code Style - `output` contains confidential data should declare `sensitive = true`
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR30 - Category: Code Style - Dealing with Deprecated `output`s
+
+Sometimes we notice that the name of certain `output` is not appropriate anymore, however, since we have to ensure forward compatibility in the same major version, it's not allowed to change the name directly. We need to move it to an independent `deprecated_outputs.tf` file, then redefine a new output in `output.tf` and make sure it's compatible everywhere else in the module.
+
+A cleanup can be performed to `deprecated_outputs.tf` and other logics related to compatibility during a major version upgrade.
+
+<br>
+
+---
+
+<br>
+
+#### ID: TFNFR31 - Category: Code Style - `locals.tf` **MUST** contain only one `locals` block
+
+All `local` **MUST** be defined in the only `locals` block in `locals.tf` file.
```

Comment:
```suggestion
All `local` **MUST** be defined in the `locals.tf` file. Multiple `locals{}` blocks are permitted.
```

Reply:
I still want to keep `locals.tf` file "pure", it should depends on `variables.tf` only, so we can run some light weight tests by using symbolic links.

I'd like to keep it open to declare `locals` right beneath `resource` and `data` blocks if a `local` referenced these blocks. WDYT?@matt-FFFFFF

Reply:
I think the use of locals next to resources should be an exception that is only used when we are considering using a fake module to test expressions in `locals.tf` - this is quite an advanced scenario. As long as we add a statement to this effect then I am happy.